File: practice/algorithms/graph_theory/even_tree/even_tree.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

int countCh(vector<set<int>> & i_adjList, int i_par)
{
    int cnt = i_adjList[i_par].size();
    if(cnt)
    {
        set<int>::const_iterator it = i_adjList[i_par].begin();
        for(; it != i_adjList[i_par].end(); it++)
        {
            cnt += countCh(i_adjList, *it);
        }
    }
    
    return cnt;
}

int main()
{    
    int n, m;
    cin >> n >> m;
    
    vector<set<int>> adjList(n + 1);
    int v1, v2;
    for (int i = 0; i < m; ++i)
    {
        cin >> v1 >> v2;
        adjList[v2].insert(v1);
    }
    
    vector<int> numCh(n + 1, 0);
    for (int i = 1; i < adjList.size(); i++)
    {
        numCh[i] = countCh(adjList, i);
    }
    
    sort(numCh.begin(), numCh.end());
    int numEdge = 0;
    for (int i = 0; i < numCh.size() - 1; ++i)
    {
        if (numCh[i] % 2 == 1)
        {
            numEdge++;
        }
    }
    
    cout << numEdge << "\n";
    
    return 0;
}

File: practice/algorithms/graph_theory/even_tree/even_tree.py
#!/usr/bin/env python

def count_ch(i_adj_list, i_par):
    cnt = len(i_adj_list[i_par])
    if cnt > 0:
        for VAL in i_adj_list[i_par]:
            cnt += count_ch(i_adj_list, VAL)
    return cnt

n, m = map(int, raw_input().strip().split())
adj_list = [set() for _ in range(n + 1)]
for _ in range(m):
    v1, v2 = map(int, raw_input().strip().split())
    adj_list[v2].add(v1)

num_ch = [0 for _ in range(n + 1)]
for POS in range(1, n + 1):
    num_ch[POS] = count_ch(adj_list, POS)
    
num_ch = sorted(num_ch)

num_edge = 0
for POS in range(n):
    if num_ch[POS] % 2 == 1:
        num_edge += 1
print num_edge

